app/src/main/java/com/example/travelapp/BinaryMinHeap.java

```diff
@@ -0,0 +1,166 @@
+ * tree corresponds to an element in the array. Taken and adapted
+ * from a Penn homework assignment.
+ * @param <Key> - The key used to determine priority in the heap.
+ *             Does not have to be distinct.
+ * @param <V> - The value of what we are adding to the heap. In
+ *           Dijkstras, this would correspond to the node number.
+ */
+
+public class BinaryMinHeap<Key extends Comparable<Key>, V> {
+
+    // arrayList can be viewed as a nearly complete binary tree
+    private ArrayList<Entry<Key, V>> arrayList;
+    // indices is a HashMap used for easier lookup of where values
+    // are in the array
+    private HashMap<V, Integer> indices;
+
+    public BinaryMinHeap() {
+        arrayList = new ArrayList<Entry<Key, V>>();
+        indices = new HashMap<V, Integer>();
+    }
+
+    public int size() {
+        return arrayList.size();
+    }
+
+    public boolean containsValue(V value) {
+        return indices.containsKey(value);
+    }
+
+    public void add(Key key, V value) {
+        if (key == null) {
+            throw new IllegalArgumentException("key is null!");
+        }
+        if (indices.containsKey(value)) {
+            throw new IllegalArgumentException("value is already in the min-heap!");
+        }
+
+        arrayList.add(new Entry<Key, V>(key, value));
+        // set index to where we just added the Entry (at the end = index size - 1 of the arrayList)
+        int index = arrayList.size() - 1;
+        // calculates the index of the parent of the current node, given the current node's index
+        Function<Integer, Integer> calculateParentIndex = (n) -> { return (n + 1) / 2 - 1; };
+        int parent = calculateParentIndex.apply(index);
+
+        indices.put(value, index);
+        placeElementAtIndex(index, parent);
+
+    }
+
+    /*
+        placeElementAtIndex is a helper method that places the element in the passed in index where
+        it should belong in the heap. called by decreaseKey and add.
+     */
+
+    private int placeElementAtIndex(int index, int parent) {
+        if (arrayList.size() == 1) {
+            return 0;
+        }
+        while (index > 0 && arrayList.get(parent).key.compareTo(arrayList.get(index).key) > 0) {
+            //exchange the entry at parent and entry at index in arrayList and in indices
+            Entry<Key, V> parentKey = arrayList.get(parent);
+            Entry<Key, V> indexKey = arrayList.get(index);
+            arrayList.set(parent, indexKey);
+            arrayList.set(index, parentKey);
+            indices.replace(indexKey.value, parent);
+            indices.replace(parentKey.value, index);
+            // updates the index value
+            index = parent;
+            // calculates the new index of the parent of the current node,
+            // given the current node's index
+            Function<Integer, Integer> calculateParentIndex = (n) -> { return (n + 1) / 2 - 1; };
```

Comment:
yeah, so if this is pulled out of individual functions, you wouldn't have to rewrite it every time, you could just call `calculateParentIndex()`